ghcjs-src/Miso/Subscription/WebSocket.hs

```diff
@@ -151,48 +129,207 @@ data SocketState
+newWebSocket_ :: MisoString -> [MisoString] -> IO (Either MisoString WebSocket_)
+{-# INLINE newWebSocket_ #-}
+newWebSocket_ url protocols = do
+  result <- newWebSocketOrError' url =<< toJSVal protocols
+  isWebSocket' result >>= \case
+    True -> pure (Right (WebSocket_ result))
+    False -> Left <$> getExceptionMessage' result
+
+-- | Create a new high-level websocket.
+newWebSocketWithProtocols
+  :: MisoString -> [MisoString] -> IO (Either MisoString WebSocket)
+{-# INLINE newWebSocketWithProtocols #-}
+newWebSocketWithProtocols url protocols = newWebSocket_ url protocols >>= \case
+  Left err -> pure (Left err)
+  Right socket_ -> Right . WebSocket url protocols <$> newMVar socket_
+
+-- | Create a new high-level websocket with an empty protocol list
+newWebSocket :: MisoString -> IO (Either MisoString WebSocket)
+{-# INLINE newWebSocket #-}
+newWebSocket = flip newWebSocketWithProtocols []
+
+-- | Close a high-level websocket.
+closeWebSocket :: WebSocket -> IO ()
+closeWebSocket ws@(WebSocket _ _ ref) = tryTakeMVar ref >>= \case
+  Nothing -> putStrLn $ concat ["WebSocket ", show ws, " already closed"]
+  Just ws_ -> closeWebSocket_ ws_
+
+-- | Close a websocket with the default code (1000)
+closeWebSocket_ :: WebSocket_ -> IO ()
+closeWebSocket_ = closeWebSocketWithCode 1000
+
+-- | Default ping config. Sends "ping" every 30 seconds and expects "pong".
+defaultPingConfig :: PingConfig
+defaultPingConfig = PingConfig 30 "ping" "pong"
+
+-- | Default websocket config.
+-- Doesn't set up a ping, and attempts to reconnect if close was abnormal.
+defaultWSConfig :: WebSocketConfig
+defaultWSConfig = WebSocketConfig {
+  websocketPing = Nothing,
+  websocketShouldReconnectOnClose = \event -> pure (not $ closedCleanly event)
+  }
+
+--------------------------------------------------------------------------------
+-- * Creating Subscriptions
+
+
+-- | WebSocket subscription with default configuration.
+websocketSub
+  :: FromJSON message
+  => WebSocket                          -- ^ Handle to the socket.
+  -> (WebSocketEvent message -> action) -- ^ Message handler.
+  -> Sub action model                   -- ^ A subscription.
+websocketSub = websocketSubWithConfig defaultWSConfig
+
+
+-- | WebSocket subscription, given a particular configuration.
+--
+-- If the socket disconnects, the config will be used to determine
+-- whether to reconnect; if so it happens seamlessly. If not, the
+-- WebSocketClosed event will be fired.
+websocketSubWithConfig
+  :: FromJSON message
+  => WebSocketConfig                    -- ^ Configuration
+  -> WebSocket                          -- ^ Handle to the socket.
+  -> (WebSocketEvent message -> action) -- ^ Event handler.
+  -> Sub action model                   -- ^ A subscription
+websocketSubWithConfig config socket handler getModel sink = do
+  let (url, protocols) = (wsURL socket, wsProtocols socket)
+
+  -- Spin off ping thread if configured
+  pingThreadId <- case websocketPing config of
```

Comment:
You can simplify this using `for`.

Reply:
simpler, but less readable? I guess I'm fine with it but sometimes using combinators obscures things.